pkg/controller/tide/tide.go

```diff
@@ -0,0 +1,729 @@
+			node.Spec.Taints[i].Key = pool.GetEvictOfflinePodTaint().Key
+			node.Spec.Taints[i].Value = pool.GetEvictOfflinePodTaint().Value
+			node.Spec.Taints[i].Effect = corev1.TaintEffect(pool.GetEvictOfflinePodTaint().Effect)
+			found = true
+		}
+	}
+	if !found {
+		node.Spec.Taints = append(node.Spec.Taints, corev1.Taint{
+			Key:    pool.GetEvictOfflinePodTaint().Key,
+			Value:  pool.GetEvictOnlinePodTaint().Value,
+			Effect: corev1.TaintEffect(pool.GetEvictOnlinePodTaint().Effect),
+		})
+	}
+	delete(node.Labels, pool.GetOfflineLabel().Key)
+	node.Labels[pool.GetOnlineLabel().Key] = pool.GetOnlineLabel().Value
+	return node
+}
+
+func (t *Tide) changeNodeToOffline(node *corev1.Node, pool NodePoolWrapper) *corev1.Node {
+	found := false
+	for i := range node.Spec.Taints {
+		if node.Spec.Taints[i].Key == pool.GetEvictOnlinePodTaint().Key || node.Spec.Taints[i].Key == pool.GetEvictOfflinePodTaint().Key {
+			node.Spec.Taints[i].Key = pool.GetEvictOnlinePodTaint().Key
+			node.Spec.Taints[i].Value = pool.GetEvictOnlinePodTaint().Value
+			node.Spec.Taints[i].Effect = corev1.TaintEffect(pool.GetEvictOnlinePodTaint().Effect)
+			found = true
+		}
+	}
+	if !found {
+		node.Spec.Taints = append(node.Spec.Taints, corev1.Taint{
+			Key:    pool.GetEvictOfflinePodTaint().Key,
+			Value:  pool.GetEvictOnlinePodTaint().Value,
+			Effect: corev1.TaintEffect(pool.GetEvictOnlinePodTaint().Effect),
+		})
+	}
+	delete(node.Labels, pool.GetOnlineLabel().Key)
+	node.Labels[pool.GetOfflineLabel().Key] = pool.GetOfflineLabel().Value
+	return node
+}
+
+func getNodeUsageWithSelector(
+	nodes simulator.ClusterSnapshot,
+	resourceNames []corev1.ResourceName,
+	selector labels.Selector,
+) ([]NodeUsage, error) {
+	var nodeUsageList []NodeUsage
+	nodeInfos, err := nodes.NodeInfos().List()
+	if err != nil {
+		return nil, err
+	}
+	for _, node := range nodeInfos {
+		var pods []*corev1.Pod
+		for i := range node.Pods {
+			pods = append(pods, node.Pods[i].Pod)
+		}
+		if !selector.Matches(labels.Set(node.Node().GetLabels())) {
+			continue
+		}
+		nodeUsageList = append(nodeUsageList, NodeUsage{
+			node:    node.Node(),
+			usage:   nodeutil.NodeUtilization(pods, resourceNames),
+			allPods: pods,
+		})
+	}
+	// nodes are sorted by usage rate
+	sort.Slice(nodeUsageList, func(i, j int) bool {
+		ti := nodeUsageList[i].usage[corev1.ResourceMemory].Value() + nodeUsageList[i].usage[corev1.ResourceCPU].MilliValue() + nodeUsageList[i].usage[corev1.ResourcePods].Value()
+		tj := nodeUsageList[j].usage[corev1.ResourceMemory].Value() + nodeUsageList[j].usage[corev1.ResourceCPU].MilliValue() + nodeUsageList[j].usage[corev1.ResourcePods].Value()
+		// extended resources
+		for name := range nodeUsageList[i].usage {
+			if !nodeutil.IsBasicResource(name) {
+				ti = ti + nodeUsageList[i].usage[name].Value()
+				tj = tj + nodeUsageList[j].usage[name].Value()
+			}
+		}
+		return ti < tj
+	})
+	return nodeUsageList, nil
+}
+
+// clearUnexpectedCNR is used to clear unexpected cnr
```

Comment:
It seems that this comment may be not relevant to this method.

Reply:
done